File: 13_Stack & Queue/E20. Valid Parentheses.java
// 20. Valid Parentheses
// https://leetcode.com/problems/valid-parentheses/

class Solution {
    public boolean isValid(String s) {
        char[] stack = new char[s.length()];
        int top = -1;
        
        for (char c : s.toCharArray()) {
            if (c == '(' || c == '[' || c == '{') {
                stack[++top] = c; // push
            } else {
                if (top == -1) return false; // empty
                char open = stack[top--];    // pop
                if (c == ')' && open != '(') return false;
                if (c == ']' && open != '[') return false;
                if (c == '}' && open != '{') return false;
            }
        }
        
        return top == -1;
    }
    
    public static void main(String[] args) {
        Solution solution = new Solution();
        String input = "()[]{}";
        boolean result = solution.isValid(input);
        System.out.println(result);
    }
}

File: JavaScript/13_Stack & Queue/E20. Valid Parentheses.js
// 20. Valid Parentheses
// https://leetcode.com/problems/valid-parentheses/

// Time Complexity: O(n)
// Space Complexity: O(n)

/**
 * Check if a string of parentheses is valid
 * @param {string} s - String containing parentheses
 * @returns {boolean} - True if valid
 */
function isValid(s) {
    const stack = [];

    for (const c of s) {
        if (c === '(' || c === '[' || c === '{') {
            stack.push(c); // push opening bracket
        } else {
            if (stack.length === 0) return false; // empty stack
            const open = stack.pop(); // pop
            if (c === ')' && open !== '(') return false;
            if (c === ']' && open !== '[') return false;
            if (c === '}' && open !== '{') return false;
        }
    }

    return stack.length === 0;
}

// Test cases
console.log(isValid("()")); // true
console.log(isValid("()[]{}")); // true
console.log(isValid("(]")); // false
console.log(isValid("([)]")); // false
console.log(isValid("{[]}")); // true
console.log(isValid("")); // true
console.log(isValid("(((")); // false